### ISSUE TITLE
Open API available?

### ISSUE THREAD

**Issue by michaelrambeau** [alexfernandez/package-quality#45]
Hello Alex,
I created an application http://bestof.js.org/ to help developers find the best components of the web platform (front and back end).
Basically it is a curated list of 500 projects on Github about js in the browser and in node.js, html, css,..
Every day the application checks variations of stars on Github to detect trends.
I would like to display more relevant information about the **quality** of the projects.
So I wonder if there is a kind of open API to access data generated by http://packagequality.com/ so that I could display the metrics you generate ?
Of course, I would add links to your application and add credits to make obvious the source of the information.
Thank you in advance for your answer and feel free to give me any feedback about bestof.js.org, especially if it is bad feedback ;)

**Comment by alexfernandez** [comment]
Hi @michaelrambeau, right now there is no API, but it would be quite easy to add a route `/api/:package` to [`bin/app.js`](https://github.com/alexfernandez/package-quality/blob/master/bin/app.js). We would be glad to merge it, so that you could just access `http://packagequality.com/api/mypackage` and get back a JSON with some info. Would you be interested in sending a PR?

**Comment by alexfernandez** [comment]
By the way, your project bestofjs.org looks great! I have noticed that there is no category for load testing, and my own project [`loadtest`](https://www.npmjs.com/package/loadtest) is not featured under "testing tools". What should I do to include it?

**Comment by michaelrambeau** [comment]
Thank you very much for your answer Alex.
OK I am going to check how package-quality works and I will send you a PR.

About adding new projects, users should be able to submit new projects, it is one of the missing features I will add soon.
`loadtest` looks good, so far there was no load test library, so I have just added it.
It will be available in the UI very soon (technically in less than 2 days because the app need at least 2 daily snapshots of Github stars to display tendencies).
I will confirm very soon.

**Comment by michaelrambeau** [comment]
Hello Alex, just to confirm that loadtest project is available in bestof.js.org, in the "test" category => http://bestof.js.org/tags/test
FYI, I am working right now on the PR about the "open API".

**Comment by alexfernandez** [comment]
That's cool, thanks! Let me know if I can help you with the PR.

**Comment by michaelrambeau** [comment]
I was checking the existing routes and it seems that what I need already exists: the `/package/:package` route already returns JSON data.

For example: http://packagequality.com/package/redux returns

```
{
	"_id": "54e2cd1ff22796cf67c857ee",
	"name": "redux",
	"source": "npm",
	"created": "2015-03-10T09:57:27+00:00",
	"lastUpdated": "2016-06-05T09:00:04+00:00",
	"nextUpdate": "2015-03-25T10:30:26+00:00",
	"timesUpdated": 2,
	"repoTotalIssues": [0.9994314951677089, 1],
	"repoOpenIssues": [1, 1],
	"repoLongOpenIssues": [1, 1],
	"versions": [0.9811320754716981, 1],
	"downloads": [0.9999997215107596, 1],
	"quality": 0.9961126584300335,
	"description": "Predictable state container for JavaScript apps"
}
```

What do you think, can I use this route ? Or is it better to create a new `/api/:package` route as discussed before ?

**Comment by alexfernandez** [comment]
Wow, I forgot. @DiegoTUI, you built this API, do you think it is usable for this purpose?

**Comment by alexfernandez** [comment]
I mean, if it contains the data you need, definitely use it, Michael :)

**Comment by DiegoTUI** [comment]
Yep!!! We do not have a documentes API, but we do have certain endpoints that we use in the packagequality.com website to get the data for a certain package. You just found the right route for you!! No need to implement the `/api/:package` then.

**Comment by michaelrambeau** [comment]
Thank you fo your answer @DiegoTUI and @alexfernandez !
I will try to include packagequality scoring in my application.
For example, for a given "tag", it could be nice to sort results by scoring, I am curious to see if some packages really stand out compared to others, I will let you know...

**Comment by michaelrambeau** [comment]
Hello Alex, I am still working to display results from packagequality.com API.

I'd like know if it's possible to find "scoped projects" like [@cycle/core](https://www.npmjs.com/package/@cycle/core) or [@pwn/is](https://www.npmjs.com/package/@pwn/is) from the UI.

![image](https://cloud.githubusercontent.com/assets/5546996/16174557/b0a0cb5a-3601-11e6-9463-cc6f976ac8aa.png)

**The API works**, for example http://packagequality.com/package/core?scope=cycle works.

... but I'd like to add links from my application to yours, when an user clicks on the quality score, that is why I am asking what is the apropriate URL to display results for a "scoped package".

Thank you!

**Comment by michaelrambeau** [comment]
Hello again Alex, according to the following comment, it does not seem that "scoped packages" are included in npm search results, so I don't think it is possible to make API calls.

So you can ignore my previous request.

https://github.com/cyclejs/core/issues/292#issuecomment-209377760

By the way , have you heard about https://npms.io/ ? The concept seems to be similar to packagequality.com.

**Comment by alexfernandez** [comment]
Hi again, we are aware that several alternatives exist to measure package quality. Our own is designed to be code-independent: it only looks at external attributes of the repo, without actually downloading it. Thus it can be applied to Maven packages, or to anything hosted in GitHub (and can be extended to others like gitorious).

The original idea was to answer the age-old question: "Yes, npm has [more packages](http://www.modulecounts.com/), but are they any good?" A language-agnostic quality check was thus devised, which also has the advantage of being more lightweight than alternatives that need to look into the code.

**Comment by michaelrambeau** [comment]
OK I see, it makes sense, thank you, that is why it's called "packagequality.com" and not "npmquality" :wink: 
And sorry for the previous request, for now I am going to ignore the "scoped packages".
I will show you the first iteration very soon...

**Comment by michaelrambeau** [comment]
Hello Alex,
packagequality.com data is now available on http://bestof.js.org/ !

![image](https://cloud.githubusercontent.com/assets/5546996/16543179/c792c48c-4103-11e6-9133-7be6871d91d3.png)

If you click on the score, it will open the package page on packagequality.com, for example http://packagequality.com/#?package=animejs

Also, for a given tag, you can sort by "quality score" if you click on the to "Quality" tab.

Let me know if you have any feedback, I can adjust things if needed, of course. Thank you in advance!

**Comment by alexfernandez** [comment]
Great news! Thanks :)

**Comment by palcarazm** [comment]
Hi,

I'm trying to use your API, but I have my XMLHttpRequest blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

Is a public API or just a private API that you don't want to expose? 

If is a public API, can you please add `Access-Control-Allow-Origin: *` header in your apache/ngix/... directive for the resource?

Obviously, I will refer your project as source of the quality score 😃 
Thanks!

**Comment by alexfernandez** [comment]
Hi @palcarazm ! Great idea, care to propose a patch? I suggest you add the header directly in the Node.js code.

**Comment by palcarazm** [comment]
> Hi @palcarazm ! Great idea, care to propose a patch? I suggest you add the header directly in the Node.js code.

Hi @alexfernandez ! 

I send a PR to add the allow-origin header. Can you do a code review and (if okay) merge and deploy?

**Comment by alexfernandez** [comment]
Hi @palcarazm, code merged and deployed. Let me know how it goes!

**Comment by palcarazm** [comment]
It's working fine with:
```javascript

fetch( "https://packagequality.com/package/bootstrap")
  .then((data) => {
    console.table(data);
  })
  .catch(()=>{});
```
